src/main/scala/is/hail/expr/Type.scala

```diff
@@ -930,6 +959,14 @@ final case class TSet(elementType: Type) extends TIterable {
   override def scalaClassTag: ClassTag[Set[AnyRef]] = classTag[Set[AnyRef]]
 }
 
+object TSet {
+  def apply(elementType: Type, elementsRequired: Boolean): TSet = {
```

Comment:
The TSet constructor (and other classes) should have an elementsRequired argument (with default = false is fine).  That will implement the "def elementsRequired" on TContainer since it is a case class and therefore the constructor arguments are implicit `val`s and `val`s can implement `def`s.

Loosely, it should like:

```
class TContainer {
  def elementsRequired: Boolean
}

case class TSet(, elementsRequired: Boolean = false) { ... }
```

and no apply on the companion object will be needed.

Reply:
I tried to do this, but I think it broke the unapply method that's sometimes used to extract 
elementType. Should I just go ahead and break that, and then fix it in all the places it's used?

Reply:
Fix it in the places that it is used.  This is an indication those uses might need to look at `elementsRequired`!